src/transaction/transaction_single_node_group_complete_manager.cpp

```diff
@@ -0,0 +1,309 @@
+// Manager of completed group on single node
+//
+
+#include "boot_sr.h"
+#include "page_buffer.h"
+#include "log_manager.h"
+#include "server_support.h"
+#include "thread_daemon.hpp"
+#include "thread_entry_task.hpp"
+#include "thread_manager.hpp"
+#include "transaction_single_node_group_complete_manager.hpp"
+
+namespace cubtx
+{
+  single_node_group_complete_manager *gl_single_node_gcm = NULL;
+  cubthread::daemon *gl_single_node_gcm_daemon = NULL;
+
+  //
+  // single_node_group_complete_task is class for master group complete daemon
+  //
+  class single_node_group_complete_task : public cubthread::entry_task
+  {
+    public:
+      /* entry_task methods */
+      void execute (cubthread::entry &thread_ref) override
+      {
+	if (!BO_IS_SERVER_RESTARTED ())
+	  {
+	    return;
+	  }
+
+	cubthread::entry *thread_p = &cubthread::get_entry ();
+	get_single_node_gcm_instance ()->do_prepare_complete (thread_p);
+      }
+  };
+
+  single_node_group_complete_manager::single_node_group_complete_manager ()
+  {
+    LSA_SET_NULL (&gl_single_node_gcm->m_latest_closed_group_start_log_lsa);
+    LSA_SET_NULL (&gl_single_node_gcm->m_latest_closed_group_end_log_lsa);
+  }
+
+  single_node_group_complete_manager::~single_node_group_complete_manager ()
+  {
+  }
+
+  //
+  // notify_log_flush notifies single node group complete.
+  //
+  void single_node_group_complete_manager::notify_log_flush_lsa (const LOG_LSA *lsa)
+  {
+    assert (lsa != NULL);
+
+    if (LSA_GE (lsa, &m_latest_closed_group_end_log_lsa))
+      {
+	cubthread::entry *thread_p = &cubthread::get_entry ();
+	do_complete (thread_p);
+      }
+  }
+
+  //
+  // get_manager_type - get manager type.
+  //
+  int single_node_group_complete_manager::get_manager_type () const
+  {
+    return LOG_TRAN_COMPLETE_MANAGER_SINGLE_NODE;
+  }
+
+  //
+  // on_register_transaction - on register transaction specific to single node.
+  //
+  void single_node_group_complete_manager::on_register_transaction ()
+  {
+    /* This function is called after adding a transaction to the current group. */
+    assert (get_current_group ().get_container ().size () >= 1);
+#if defined (SERVER_MODE)
+    if (is_latest_closed_group_completed ())
+      {
+	/* We can move the next calls outside of mutex. */
+	if (can_wakeup_group_complete_daemon (true)
+	    || pgbuf_has_perm_pages_fixed (&cubthread::get_entry ()))
+	  {
+	    /* This means that GC thread didn't start yet group close. */
+	    gl_single_node_gcm_daemon->wakeup ();
+	  }
+      }
+    else if (!is_latest_closed_group_complete_started ()
+	     && is_latest_closed_group_prepared_for_complete ())
+      {
+	/* Be sure that log flush knows that GC thread waits for it. */
+	log_wakeup_log_flush_daemon ();
+      }
+#endif
+  }
+
+#if defined (SERVER_MODE)
+  //
+  // can_wakeup_group_complete_daemon - true, if can wakeup group complete daemon
+  //
+  bool single_node_group_complete_manager::can_wakeup_group_complete_daemon (bool inc_gc_request_count)
```

Comment:
bool inc_gc_request_count : an enum { DONT_USE_STATS, USE_STATS}  instead of true/false would help readability.